build/rpmfc.c

```diff
@@ -1184,22 +1184,42 @@ static rpmRC rpmfcApplyInternal(rpmfc fc)
 static int initAttrs(rpmfc fc)
 {
     ARGV_t files = NULL;
+    int nfiles = 0;
     char * attrPath = rpmExpand("%{_fileattrsdir}/*.attr", NULL);
-    int nattrs = 0;
+    ARGV_t all_attrs = argvNew();
 
-    /* Discover known attributes from pathnames + initialize them */
+    /* Discover known attributes from pathnames */
     if (rpmGlob(attrPath, NULL, &files) == 0) {
-	nattrs = argvCount(files);
-	fc->atypes = xcalloc(nattrs + 1, sizeof(*fc->atypes));
-	for (int i = 0; i < nattrs; i++) {
-	    char *bn = basename(files[i]);
-	    bn[strlen(bn)-strlen(".attr")] = '\0';
-	    fc->atypes[i] = rpmfcAttrNew(bn);
-	}
-	fc->atypes[nattrs] = NULL;
-	argvFree(files);
+	nfiles = argvCount(files);
+    }
+    for (int i = 0; i < nfiles; i++) {
+	char *bn = basename(files[i]);
+	bn[strlen(bn)-strlen(".attr")] = '\0';
+	argvAddUniq(&all_attrs, bn);
```

Comment:
The files coming out of rpmGlob() are unique already, you don't need argvAddUniq() here. What I meant is that you could simply use the argv coming from rpmGlob() and argvAddUniq() the local stuff to that.

Reply:
Yes, but they are file names and not just attribute names. An while they are unique because we hard coded the glob, making the location configurable may turn up the same names from different locations. That's why I made the code a bit more defensive than it needs to be. Also I don't even want to think what rpmGlob returns if it doesn't find anything. Overall I don't think this is going to get that much simpler in the end.

Reply:
Doh, I was somehow thinking rpmGlob() returned basenames there but, but (of course) it doesn't. So scratch that part, sorry.

The location is configurable but it's still always just a single directory, it cannot possibly have multiple files by the same name. It's not that argvAddUniq() is *wrong*, it just gives the reader the wrong impression basically. Similarly moving this .attr splitting loop out of the rpmGlob() if-block seems strange, because there's only ever anything to do on rpmGlob() success. So that's kinda backwards to what you say about not wanting to think about the case where it doesn't find anything because that's already handled, and just obfuscates the diff for no good reason. It's always better to let the actual change stand out in the diff when reasonably possible.

This is the *actual* change to the first part here. nattr gets recalculated later on, but most of the time it's already correct from here:

```
     /* Discover known attributes from pathnames + initialize them */
     if (rpmGlob(attrPath, NULL, &files) == 0) {
        nattrs = argvCount(files);
-       fc->atypes = xcalloc(nattrs + 1, sizeof(*fc->atypes));
        for (int i = 0; i < nattrs; i++) {
            char *bn = basename(files[i]);
            bn[strlen(bn)-strlen(".attr")] = '\0';
-           fc->atypes[i] = rpmfcAttrNew(bn);
+           argvAdd(&all_attrs, bn);
        }
-       fc->atypes[nattrs] = NULL;
        argvFree(files);
     }
+
+    /* Get file attributes from _local_file_attrs macro */
+    /* ... /

```
(actually one could just pass &nattrs to rpmGlob() and avoid the argvCount(), but that too would kinda obfuscate the diff)


Patch cleanliness aside, this all needs to be just one commit. It's a single new feature with its tests, we are not interested in PR review in the commit history.

Reply:
Sure. There also is a "squash and merge" button somewhere  down there...

Reply:
I realised the really right way to do this is to refactor the discovery/init split to a separate commit, and then the local_attr thing doesn't need to change anything at all, only add itself.
Having asked this to be one commit already I couldn't very well ask you to do this, so here goes: https://github.com/rpm-software-management/rpm/pull/2911